feat(ONB-201): Keyword Generator implemented

ufukagaya opened the pull request:
- Implemented Keyword Generator page at address /keyword-generator.
- Keywords are sorted according to their frequency of use in the app description and distinguished with colors.
- Keywords clicked on are highlighted in the app description provided.

sevvalkaplannn reviewed (changes requested):
https://github.com/user-attachments/assets/bdd2219e-75b4-42b7-b590-95c8b72cbc68

hi ufuk, 

1- when we generate the keywords in the text input we are not able to edit the input unless we click the clear button, it is a way that doing it so it is not wrong way to do it, but for the user experience it is not the best way to do it, we should be able to delete or edit the text input whenever we want, we shouldn't need to click the clear button, because maybe user wanna add a new keyword or made a typo but now when we want to make changes we need to clear all of the text, it is not proper for input component. Can you change it?

2- can you make a home page with a button that directs to the relevant page so that we don't have to go through the URL for keyword generator page?

3- you shouldn't use bootstrap for css, you should use css flexbox, for that you can read those documentations

- https://www.w3schools.com/css/css3_flexbox.asp
- https://css-tricks.com/snippets/css/a-guide-to-flexbox/

sevvalkaplannn reviewed (changes requested):
ufuk hi, you shouldn't add the node modules into the pr, please go to the master and copy the .gitignore and paste it in to yours

sevvalkaplannn reviewed (changes requested):
hi ufuk

1- you don't need to make two displays for input, below input is enough for our case, but the css is not aligned. Can you write it with better alignments, the input overflows from your borders. Please check the example input in the documentation.

<img width="1348" alt="Screenshot 2025-07-02 at 23 25 53" src="https://github.com/user-attachments/assets/d6c00a7f-886e-4d05-aecc-8797e034cb81" />

2- when writing css classes you need to name it better, for example for .home-button class you should name it like ma-home-button, you should add ma-.. for the naming convention (ma means mobileaction for writing company based code), and the names of the CSS classes should be compatible with each other and when we see the class name, it should give information about where it is used.

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
 <script setup>
```
in the script setup part there is a rule that you should follow for implementation order it should follow the below ruleset;

1- Props and Emits
2- Ref and Reactive data
3- Computed property
4-Method
5-Function description
6- Lifecycle hooks

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+</template>
+
+<style scoped>
+.ma-home-button {
+    position: absolute;
+    top: 1rem;
+    left: 1rem;
+    text-decoration: none;
+}
+.ma-container {
+    margin-top: 5rem;
+    max-width: 100vw;
+    overflow-x: hidden;
+}
```
Can you wrap the CSS classes within each other according to the hierarchy between the divs, like the example below, this will increase readability. 

`
.ma-container {
    margin-top: 5rem;
    max-width: 100vw;
    overflow-x: hidden;
    .ma-home-button {
       position: absolute;
       top: 1rem;
       left: 1rem;
       text-decoration: none;
    }
}`

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+    text-align: center;
+    white-space: nowrap;
+    vertical-align: middle;
+    -webkit-user-select: none;
+    -moz-user-select: none;
+    -ms-user-select: none;
+    user-select: none;
```
I think you are writing unnecessary css's can you check the css's and delete the ones that you do not need.

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+    font-weight: 600;
+    line-height: 1;
+    text-align: center;
+    white-space: nowrap;
+    vertical-align: baseline;
```
for example you don't need them

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+    <div class="ma-container">
+        <router-link to="/" class="ma-btn ma-btn-primary ma-home-button">Home</router-link>
+        <div class="ma-row">
+            <div class="ma-col-left">
+                <div class="ma-card">
+                    <div class="ma-card-body">
```
you are wrapping the same code block with 3 different div's, you don't need them you can just have one and you can collect the css's in one class.

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+                        </ul>
+                    </div>
+                    <div class="ma-card-body">
+                        <div v-if="currentKeywords.length">
+                            <div class="ma-keywords-container">
```
in here too, you don't need to wrap the same code block with two different divs. can you check the whole code for removing the unnecessary duplicated divs.

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
-        <div class="ma-header">
-            <span>Keyword Generator</span>
+    <div class="ma-container">
+        <router-link to="/" class="ma-btn ma-btn-primary ma-home-button">Home</router-link>
```
you already have src/components/TheNav.vue component, you can use that instead of writing this part.

sevvalkaplannn commented on src/components/TheNav.vue:
```diff
@@ -1,16 +1,23 @@
+        <router-link to="/keyword-generator" class="ma-nav-link">Keyword Generator</router-link>
     </div>
 </template>
 
 <style>
- .ma-router-item {
-   display: flex;
-   flex-direction: row;
-   justify-content: space-evenly;
-   margin: 15px;
- }
+.ma-nav-links-container {
+    display: flex;
+    flex-direction: row;
+    justify-content: space-evenly;
+    margin: 15px;
+}
+.ma-nav-link {
+    text-decoration: none;
+    color: #007bff;
+    padding: 0.5rem 1rem;
+}
+.ma-nav-link:hover {
+    text-decoration: underline;
+}
 </style>
```
you updated the TheNav.vue component but you are not using it anywhere

sevvalkaplannn commented on src/App.vue:
```diff
@@ -1,13 +1,33 @@
 
-
+<style>
+.ma-app-container {
+    display: flex;
+    flex-direction: column;
+    min-height: 100vh;
+}
+.ma-main-content {
+    flex-grow: 1;
+    display: flex;
+    justify-content: center;
+    align-items: flex-start;
+    padding: 20px;
+}
+.ma-page-wrapper {
+    width: 100%;
+    max-width: 1200px;
+}
+</style>
```
You can wrap the CSS classes within each other according to the hierarchy between the divs, like the example below, this will increase readability

`.ma-app-container {
    .....
    .ma-main-content {
        .....
        .ma-page-wrapper {
           ......
        }
     }
}
`

sevvalkaplannn commented on src/pages/index.vue:
```diff
@@ -1,10 +1,52 @@
+</template>
+
+<style scoped>
+.ma-home-container {
```
please follow the instruction that I wrote about the css class hierarchy in the above and below comments.

sevvalkaplannn commented on src/pages/index.vue:
```diff
@@ -1,10 +1,52 @@
+    text-align: center;
+    white-space: nowrap;
+    vertical-align: middle;
+    -webkit-user-select: none;
+    -moz-user-select: none;
+    -ms-user-select: none;
```
do you need them? please check the whole css codes and delete the unused ones.

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+const activeTab = ref('1-gram');
+
+const textareaRef = ref(null);
+function autoResize() {
+    nextTick(() => {
+        if (textareaRef.value) {
+            textareaRef.value.style.height = 'auto';
+            textareaRef.value.style.height = textareaRef.value.scrollHeight + 'px';
+        }
+    });
+}
+
+onMounted(autoResize);
+watch(userInput, autoResize);
```
you don't need that function, you can just give a min-height into your input class, it will visible up to the min height limit, then there is no problem with scrolling. You don't need to write js codes for that case, just give min-height

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+                            placeholder="Paste your app description here..."
+                            ref="textareaRef"
+                            @input="autoResize"
+                            style="resize:none;"
```
you don't need to add style in here, in your css classes you can write them

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+onMounted(autoResize);
+watch(userInput, autoResize);
+
+function generateKeywords() {
+    const cleaned = cleanDescription(userInput.value);
+    const words = cleaned.split(/\s+/).filter(word => word);
+
+    generatedKeywords.value = {
+        '1-gram': generateNGrams(words, 1),
+        '2-gram': generateNGrams(words, 2),
+        '3-gram': generateNGrams(words, 3),
+    };
+}
```
can you get the words then map them according their n-grams, instead of doing it for 3 of them.

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,254 @@
+                    <div class="ma-card-header">
+                        <ul class="ma-nav">
+                            <li class="ma-nav-item" v-for="(keywords, ngram) in generatedKeywords" :key="ngram">
+                                <a class="ma-nav-link" :class="{ active: activeTab === ngram }" @click="activeTab = ngram" href="#">
+                                    {{ ngram }} ({{ keywords.length }})
+                                </a>
```
for example this is a better feedback, I missed that.

ufukagaya commented:
- Refactored and simplified the KeywordGenerator.vue component:
- Removed unnecessary wrapper divs and redundant classes.
- Ensured all class names use the ma- prefix and are written in a single line in CSS.
- Moved button styles  from the global CSS file into the component’s scoped style block.
- Updated the <script setup> section to follow a clear order: refs, computed, methods, helpers.
- Optimized n-gram generation to process all n-grams in a single loop.
- Ensured cleanDescription uses utilities from cleanupResources.js.
- Removed all JavaScript related to textarea auto-resizing; now uses only CSS with min-height and overflow-y: auto.
-Updated App.vue, index.vue, TheNav.vue components to use <style lang="scss" scoped> and single-line class definitions for consistency.

sevvalkaplannn reviewed (changes requested):
<img width="847" alt="Screenshot 2025-07-03 at 16 57 28" src="https://github.com/user-attachments/assets/5c38aa2d-f017-4373-8801-ee3949178a03" />

<img width="527" alt="Screenshot 2025-07-03 at 16 57 48" src="https://github.com/user-attachments/assets/20194eea-f51c-4c74-b4e2-ba4c5d8e75bd" />

you have errors on your src/pages/keyword-generator/KeywordGenerator.vue file ufuk can you fix them? 

and in your own ide doesn't it show you those errors? 

2- We misunderstood about the feedback about css classes, I did not mention that the css codes should be single-line, (it will be written as single-line when using tailwind in the later stages, but not for plain css). What I wanted to say is for example in your src/pages/keyword-generator/KeywordGenerator.vue file you have div which has .ma-container class, and it wraps the div with ma-card-content class, so in your css you can write them nested too. I showed you in below picture for just one. 

<img width="591" alt="Screenshot 2025-07-03 at 17 05 08" src="https://github.com/user-attachments/assets/550dde3c-7648-4380-8631-f2e3ff33fba8" />

<img width="582" alt="Screenshot 2025-07-03 at 17 04 51" src="https://github.com/user-attachments/assets/7135f772-a66f-4a80-aa68-92b967c28def" />

sevvalkaplannn commented on src/pages/keyword-generator/KeywordGenerator.vue:
```diff
@@ -1,9 +1,211 @@
 <script setup>
+import { ref, computed } from 'vue';
+import { regex, splitRegex } from '../../cleanupResources.js';
```
ufuk in src/utils/CleanDescription.js we are using it in a function, you should call that function, not them

ufukagaya commented:
- Refactored Nav.vue usage.
- Deleted and simplified unnecessary classes.

ufukagaya commented:
- Removed style usage.
- Improved button visions

sevvalkaplannn commented on src/components/TheNav.vue:
```diff
@@ -1,16 +1,20 @@
-            Keyword Density
-        </router-link>
-    </div>
+    <button class="ma-btn ma-btn-secondary ma-home-fixed" @click="goHome">Home</button>
```
you are not using the extra class in here
```suggestion
    <button class="ma-home-fixed" @click="goHome">Home</button>
```

sevvalkaplannn commented on src/components/TheNav.vue:
```diff
@@ -1,16 +1,20 @@
-   justify-content: space-evenly;
-   margin: 15px;
- }
+<style scoped>
```
```suggestion
<style lang="scss" scoped>
```
